### ISSUE TITLE
grasping pose approach (idea n1)

### ISSUE THREAD

**Issue by giuliavezzani** [tacman-fp7/handover#15]
Start design grasping pose in case we are exploiting grasping pose idea n1

**Comment by giuliavezzani** [comment]
Code structure:

- [x] **_apriori_ poses** are read from the model
- [x] the **estimated object pose** can be read from a text file or received from [**localizer module**](https://github.com/tacman-fp7/handover/tree/master/in-hand-localization/src/localizer)
- [x] the _apriori_ poses are **rotated and translated** so that to be overlapped to the object model in the correct pose
- [x] the new poses are **shown** on the robot **cameras**
- [x] the new poses have to be evaluated for the grasping of the second hand

**Comment by giuliavezzani** [comment]
How to **select** a good pose among all the _apriori_ poses?

1. **manipulability index**
2. **distance** from joints **limits**
3. **first hand avoidance**. 
         - simple idea: assume to have symmetric _apriori_ poses (see [#14](https://github.com/tacman-fp7/handover/issues/14#issuecomment-262912820) and the following figure).  Consider the poses grouped in 4 couples: 2 on the top (3,4), 2 on the right (5,6), 2 on the bottom (7,8) and 2 on the left (1,2) of the object. Considering the object pose in the robot hand, we could **discard those couples that are closer to the robot hand**.

 For example, in the following picture
![poses2 1](https://cloud.githubusercontent.com/assets/9597070/20665541/61d18b66-b560-11e6-8029-a63685b67b6a.png)
couples **(1,2)** and **(7,8)** would be **discarded** since they are closer to the robot hand.

**Comment by giuliavezzani** [comment]
The  pipeline:
- [**point cloud acquisition and  filtering**](https://github.com/tacman-fp7/handover/tree/master/in-hand-localization/src/in-hand-segmentation)
- [**object localization**](https://github.com/tacman-fp7/handover/tree/master/in-hand-localization/src/localizer)
- [**_apriori_ poses visualization**](https://github.com/tacman-fp7/handover/tree/master/grasping-pose-selection)

has been tested successfully on the robot ([c875309](https://github.com/tacman-fp7/handover/commit/c875309be710b90b9cd087bf45f13c6810e19532)).

Hereafter, the obtained results are shown and some comments are provided.

### Results

**Filtered point cloud**. The light blue points are the points filtered with **coarse filter**, **hand filter** and **ellipse filter**:

![filtered-points](https://cloud.githubusercontent.com/assets/9597070/20748543/e2dacf62-b6ee-11e6-8fc7-db9453d06be0.png)

**Estimated object pose**. The object model (in gray) is fit into the filtered point cloud:

![estimated_pose](https://cloud.githubusercontent.com/assets/9597070/20748598/1db1972e-b6ef-11e6-86ea-354da3ce0c05.png)

**Pose visualization**: the 8 _apriori_ poses on the localized object model are shown on the robot camera view:

![poses](https://cloud.githubusercontent.com/assets/9597070/20748672/63ac6434-b6ef-11e6-9016-42117d2b3eb4.png)

### Comments

Pros :smiley: 
- if the point cloud is good, the **localizer works** pretty well with a simple object like the box

Cons :persevere: 
- the **quality of the point cloud**, is affected by many elements:
       - light condition (environment)
       -  SFM calibration (calibration can solve problems)
       -  object color (caused by gray filter, grayish object should be avoided) 
       -  object position in the robot hand (caused by ellipse filter, smaller object are better)

**Comment by lornat75** [comment]
Consider that some iCub are not black. For example the Darmstadt robot is dark blue. Therefore on the Darmstadt robot non colored objects may work and we will have to avoid objects that are blue instead. Just to keep in mind, for the moment this is not the major contribution of the work and can be simplified.

**Comment by giuliavezzani** [comment]
Aren't the palm and the fingers of the hand always the same? For example, the Blacky and the Purply have the same hand colors.

**Comment by giuliavezzani** [comment]
Every **pose** is represented by an **index**.
The starting indices are 0.0. Then:

* the **distance** of each pose from the robot **hand** position is computed and each index is decreased proportionally to such a distance.
* then, a possible **q vector** is computed by using the inverse kinematic for reaching the desired values. **NB**: in case they are not reachable, a different pose is taken into account in the q vector. For this reason, the **error** between the **ideal desired pose** and the **pose obtained by the inverse kinematics** is computed. Then, the **manipulability** of q together with the error is used to update the pose indices.

An example of such an approach is shown in the following figure.
The first robot hand (right) is ideally located on the bottom right corner of the object. The selected pose (number 4), together with the pose indices values (in red the lower, in green the higher ones) seems to be consistent with the proposed scenario.

**NB** The color of the pose 0 should be "more red". Some tests on a real scenario (on the robot possibly) will help to test this approach.

![poses](https://cloud.githubusercontent.com/assets/9597070/20891205/0b1a6568-bb0a-11e6-8102-f470d6571311.png)

**Comment by giuliavezzani** [comment]
The code has been tested on the robot [[e595a62](https://github.com/tacman-fp7/handover/commit/e595a62ce3a5a9013df49d5cdc583bae28f89b4a)].

**Object localization**

![localization](https://cloud.githubusercontent.com/assets/9597070/21003778/701153f8-bd2d-11e6-87ad-c3937ffd6167.png)

See discussion in the [previous comment](https://github.com/tacman-fp7/handover/issues/15#issuecomment-263835179). 

**Selected poses**
The manipulability index is periodically computed for each pose, after solving every time the inverse kinematics. In this way, two possible poses are selected:
* pose 2:

![pose2](https://cloud.githubusercontent.com/assets/9597070/21004090/f4b942c2-bd2e-11e6-875f-64bb24488347.png)

* pose 6:

![pose6](https://cloud.githubusercontent.com/assets/9597070/21003876/e5263ba4-bd2d-11e6-8cfd-1b2149164183.png)

**Pros** :smiley:
* the selected poses are **reachable** by the right hand
* the poses **colors** (and then the indices values) correctly **distinguish** between reachable and not-reachable poses. For every couple (0,1), (2,3), (4,5), (6,7),  one of them is "reddish" and the other "greenish", according to the reachability of the orientation.

**Cons** :persevere: 
* the results of the inverse kinematics are **not constant**. The selected pose changes very often
* sometimes, the **errors** between the desired pose and the pose obtained by inverting the kinematics are **large** and, therefore,  the selected pose is wrong

In conclusion, even if we are on the right track, manipulability computation needs to be investigated.

**Comment by pattacini** [comment]
Well done @giuliavezzani ✨ 

Are you sure that **pose 2** is reachable in that configuration? I don't know how the first hand is in the space, but assuming it in the configuration I saw in the past days, it seems to me that **pose 2** is barely reachable by hand 2. I might be mistaken though.

Results that are not consistent over time are somehow expected, given that we deal with a **redundant problem**. There are several ways to constrain the optimizer to refine this variability.

What I don't understand here is if the optimizer is prone to fail for desired poses that you proved beforehand are attainable.

**Comment by giuliavezzani** [comment]
I solved the mistake :smiley:
I didn't actually enable the **torso joints** (error in my code) and the poses computed by solving the inverse kinematics (called in the iCub documentation `xd_hat`, and `od_hat`) were very **different** from the desired ones (`xd`, and `od`).

I tested the new code on the simulator, replicating the scenario of the previous [comment](https://github.com/tacman-fp7/handover/issues/15#issuecomment-265692371). 
Hereafter, you can see the selected pose and all the other poses colored according to their reachability:

![test-pose](https://cloud.githubusercontent.com/assets/9597070/21013598/06057984-bd5b-11e6-8614-155f958e7fe7.png)

With the correct computation of `xd_hat` and `od_hat`, the **_greenish_ poses have zero position errors** (`xd - xd_hat`) and **low orientation errors** (`od-od_hat`). The **_reddish_ poses** have instead the same position error but **higher orientation errors**.

In this scenario, the **only selected pose** is pose number 5, since the correct solution of the inverse kinematics provides a better index for pose 5 with respect to pose 0. 
In particular, I show on the camera both the pose given by `xd_hat` and `od_hat`  and the pose given by `xd` and `od`. They are exactly the same. 

Furthermore, I also added a **control by the user** on pose selection procedure. When a pose is selected, the poses indices are not computed any more. The inverse kinematics is solved again, and then the poses indices and the selected pose are updated, only if the user is not satisfied with this pose and asks for a new pose.

**Comment by pattacini** [comment]
Great debugging!

**Comment by pattacini** [comment]
Well, hold on, you cannot really use the torso, because by doing so the first hand will start moving as well...

To ease the handover we should move the first hand towards a more comfortable configuration.

**Comment by pattacini** [comment]
The fact is that, with only **7 DOF**, the dexterous workspace is fairly limited, unfortunately.

This leads me to think that it's likely we'll need the **two hands approach**...

cc @massimoregoli

**Comment by giuliavezzani** [comment]
In order to check if the **two hands approach** is strictly necessary, I performed some further tests.
In particular, I collected a new **data set** putting the _first hand_ as much as possible near to the  **workspace** of _second hand_. 

 In the following example, the _first hand_ is the left hand. The object is located at **y=0.0 - y=0.1**, so that to be in the workspace of the _second hand_, i.e. right hand in this example:

![robot-frame](https://cloud.githubusercontent.com/assets/9597070/21519467/97904d92-cced-11e6-8eeb-6333492d98a6.png)

Hereafter, I report the acquired point cloud, the object models in the estimated poses and the pose selected with the **torso DOFs enabled** and with the **torso DOFs blocked**. 

### **test no.1**

* point cloud and estimated pose:

![testno1](https://cloud.githubusercontent.com/assets/9597070/21519745/d434de28-ccef-11e6-92ab-3698b1b5ac42.png)

* pose selected with **torso DOFs on**:

![torso_on-crop](https://cloud.githubusercontent.com/assets/9597070/21519762/f2ac719a-ccef-11e6-8ee4-b7c2e13d3746.png)

* pose selected with **torso DOFs off** (the selected pose is the number 2, but the actual pose computed with the kinematics is shifted, since pose 2 is not reachable without enabling the torso)

![torso_off-crop](https://cloud.githubusercontent.com/assets/9597070/21519779/11f59b80-ccf0-11e6-840e-c728315005d4.png)

### **test no. 2**

* point cloud and estimated pose:

![testno2](https://cloud.githubusercontent.com/assets/9597070/21519841/6e203370-ccf0-11e6-84df-15e85f03b46c.png)

* pose selected with **torso DOFs on**:

![torso_on -crop](https://cloud.githubusercontent.com/assets/9597070/21519868/9c605c42-ccf0-11e6-80d3-5f01585faa8f.png)

* pose selected with **torso DOFs off**:

![torso_off-crop](https://cloud.githubusercontent.com/assets/9597070/21519886/b2c65fe0-ccf0-11e6-846d-25034f6421f2.png)

### test no. 3

* point cloud and estimated pose:

![testno3](https://cloud.githubusercontent.com/assets/9597070/21519910/d667e3ec-ccf0-11e6-8919-79b1aceb8e59.png)

* pose selected with **torso DOFs on**:

![torso_on-crop](https://cloud.githubusercontent.com/assets/9597070/21519919/e218a62c-ccf0-11e6-9b93-b3fec0bd5358.png)

* pose selected with **torso DOFs off**:

![torso_off crop](https://cloud.githubusercontent.com/assets/9597070/21519926/ebce8ccc-ccf0-11e6-9c46-da76af8ddefd.png)

Test no. 1, no. 2, no.3 show the pose selected in 3 different poses, respectively nearer and nearer to the workspace of the second hand (right in this case). However, a good solution is not found without enabling the torso DOFs.
Hence, the **two hands approach is striclty necessary**.
